base/abstractarray.jl

```diff
@@ -97,51 +123,75 @@ linearindexing(::LinearIndexing, ::LinearIndexing) = LinearSlow()
 end
 
 # check along a single dimension
-checkbounds(::Type{Bool}, sz::Integer, i) = throw(ArgumentError("unable to check bounds for indices of type $(typeof(i))"))
-checkbounds(::Type{Bool}, sz::Integer, i::Real) = 1 <= i <= sz
-checkbounds(::Type{Bool}, sz::Integer, ::Colon) = true
-function checkbounds(::Type{Bool}, sz::Integer, r::Range)
+checkindex(::Type{Bool}, inds::UnitRange, i) = throw(ArgumentError("unable to check bounds for indices of type $(typeof(i))"))
+checkindex(::Type{Bool}, inds::UnitRange, i::Real) = (first(inds) <= i) & (i <= last(inds))
+checkindex(::Type{Bool}, inds::UnitRange, ::Colon) = true
+function checkindex(::Type{Bool}, inds::UnitRange, r::Range)
     @_propagate_inbounds_meta
-    isempty(r) | (checkbounds(Bool, sz, first(r)) & checkbounds(Bool, sz, last(r)))
+    isempty(r) | (checkindex(Bool, inds, first(r)) & checkindex(Bool, inds, last(r)))
 end
-checkbounds{N}(::Type{Bool}, sz::Integer, I::AbstractArray{Bool,N}) = N == 1 && length(I) == sz
-function checkbounds(::Type{Bool}, sz::Integer, I::AbstractArray)
+checkindex{N}(::Type{Bool}, indx::UnitRange, I::AbstractArray{Bool,N}) = N == 1 && indx == indices(I,1)
+function checkindex(::Type{Bool}, inds::UnitRange, I::AbstractArray)
     @_inline_meta
     b = true
     for i in I
-        b &= checkbounds(Bool, sz, i)
+        b &= checkindex(Bool, inds, i)
     end
     b
 end
 
-# check all dimensions
-function checkbounds{N,T}(::Type{Bool}, sz::NTuple{N,Integer}, I1::T, I...)
+# check all indices/dimensions
+function checkbounds(::Type{Bool}, A::AbstractArray, I...)
     @_inline_meta
-    checkbounds(Bool, sz[1], I1) & checkbounds(Bool, tail(sz), I...)
+    # checked::NTuple{M} means we have checked dimensions 1:M-1, now
+    # need to check dimension M. checked[M] indicates whether all the
+    # previous ones are in-bounds.
+    # By growing checked, it allows us to test whether we've processed
+    # the same number of dimensions as the array, even while
+    # supporting CartesianIndex
+    checked = (true,)
+    _chkbnds(A, (true,), I...)
+end
+# Single logical array indexing:
+_chkbnds(A::AbstractArray, ::NTuple{1,Bool}, I::AbstractArray{Bool}) = indices(A) == indices(I)
```

Comment:
My intention is that `indices` always returns a `Tuple{Vararg{UnitRange{Int}}}`, so I don't think it has to implement comparators.